### PR TITLE
Check for fused kernel before inplace update

### LINKED ISSUE #120217: input buffer sometimes incorrectly marked as inplace when it's not in Inductor
### 🚀 The feature, motivation and pitch

```
import torch
import torch.nn as nn

D = 4096
B = 2048*32
inp = torch.randn(B, D, dtype=torch.float16)
gO = torch.randn(B, D, dtype=torch.float16)
mod = nn.LayerNorm([D], dtype=torch.float16)

comp_mod = torch.compile(mod, dynamic=False)
print(D)
comp_mod(inp)
```

Generates https://pastebin.com/cJZbb5VY

In particular, ` tl.store(in_out_ptr0 + (x0), tmp10, None)` despite the fact that `in_out_ptr0` is never read from.

I suspect that in certain cases, when we fuse scheduler nodes reads that should be dropped aren't dropped?

cc: @peterbell10 @eellison @shunti

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @msaroufim @zou3519 @aakhundov

### PR DESCRIPTION
**Pull request by exclamaforte**
Summary:
Given an op, with a pair (output buffer, input buffer) from that op, we consider marking the output buffer as inline. However, if the parent of input buffer and the current op are going to be fused, then we don't want to mark the output buffer as inline. This change checks that criterion, and skips inlining if it is so.

Test Plan:
New unit test "layer_norm_should_not_inplace" runs LayerNorm and checks for no "in_out" pointers.

Fixes #120217

Here's a diagram of the issue:
Inline+Fusion

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -377,7 +377,7 @@ def can_inplace(self, read_dep: dependencies.Dep) -> bool:
     def has_side_effects(self) -> bool:
         return False
 
-    def decide_inplace_update(self) -> None:
+    def decide_inplace_update(self, fused_nodes: Optional[Set[str]] = None) -> None:
```
Comment (eellison) [inline torch/_inductor/scheduler.py:380]: We don't need to pass this through. You can access the fused node through: `self.scheduler.name_to_fused_node`

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -419,6 +419,11 @@ def decide_inplace_update(self) -> None:
                     and V.graph.wrapper_code.can_reuse(input_buf, self)
                     and not isinstance(input_buf.defining_op, NopKernelSchedulerNode)
                 ):
+                    # If the writers of input_buf are in the same FusedSchedulerNode as the current op, then there is
+                    # no need to inline.
```
Comment (eellison) [inline torch/_inductor/scheduler.py:423]: nit: /s/inline/inplace.

**Review by eellison — COMMENTED** [review]
looks good, one simplification on function calling convention

**Review by eellison — APPROVED** [review]
